### ISSUE TITLE
Protocol/Guidance on calls to standard library functions

### ISSUE THREAD

**Issue by hs-apotell** [copperspice/copperspice#182]
CopperSpice replaced all Qt containers to use standard containers i.e. containers like `QVector` is merely a wrapper around `std::vector`. That's awesome but unfortunately, this change has raised some concerns when porting existing libraries that work with Qt but not with CopperSpice. More specifically, on Windows platform, calls to `std::vector::data` or `std::vector::begin` when the container is empty throws exceptions (or asserts in debug builds).

Are there specific protocol or guidance on how to handle these issues? Are these to be fixed in CopperSpice so as to avoid such calls to standard library (to mimic what happens in Qt) or are these to be fixed in external libraries to not even bother calling CopperSpice functions. There has to be a conditional check someplace and we are looking for some consistency.

The specific problem we are running into at the moment is to do with calls to QPainter::drawXXX where the primitive to be drawn is empty (like an empty polygon or empty line list). The wrappers functions call `constData` which throws the exceptions.

Appreciate your continued support.

**Comment by bgeller** [comment]
The C++ standard library containers are well proven and are the industry standard. Having custom containers is error prone and most everyone will get it wrong in slightly different ways. Our decision to wrap the STL containers using composition was well thought out and solved multiple on going problems. One such example was the old implementation of QList, which is now an std::dequeue. 

In CopperSpice we have maintained the Qt API and also added the STL API to expand flexibility. Another advantage we gained was the ability to use the STL algorithms since our containers now conform to the underlying C++ requirements. Splitting QMap (and QHash) into QMap and QMultiMap makes it clear how the container treats duplicate keys. Adding the Compare template parameter improves the container even further. 

Your question was well thought out and is very valuable. However, keep in mind our goal is to transition CopperSpice to be a set of libraries which play nicely with C++.  CopperSpice does not need to be "fixed" since we are using the standard library.  Our solutions will always be to move towards the C++ standard rather than perpetuate incorrect behavior for perceived compatibility. 

The real question we want to address is are you seeing an assertion or an exception? Does this happen for only a QVector?  Does this occur on all platforms or just MSVC ( maybe debug only )?  Please know the CS team highly values your suggestions and observations, so thank you.  

Now, lets get busy.  We are going to see if we can reproduce this with an empty polygon and improve QVector.  

Barbara

**Comment by hs-apotell** [comment]
No arguments on switching to using standard containers. I would have preferred a _is-a relationship_ rather than a _has-a relationship_ but I will leave that discussion for some other day. I am sure there are pros and cons for both approaches.

To stay with the current topic of discussion - AFAIK, the problem is specific to windows platform and CL compiler. I don't know of any other compiler implementing iterator level debugging as discussed [here](https://docs.microsoft.com/en-us/cpp/standard-library/iterator-debug-level?view=msvc-160). These are assertions that trigger (enabled, by default, only on debug configurations) on _bad_/_wrong_ calls on the standard containers. These can be circumvented by a few defines at compile time but are nice feature to have for minimal checks on usage.

> an assertion or an exception

It's an assertion. Here's the relevant callstack -
```
CsGui.dll!std::vector<QPointF,std::allocator<QPointF>>::operator[](const unsigned __int64 _Pos=0) Line 1510
CsGui.dll!QVector<QPointF>::constData() Line 140
CsGui.dll!QPainter::drawPolygon(const QPolygonF & polygon={...}, Qt::FillRule fillRule=OddEvenFill) Line 540
```

> Does this happen for only a QVector?

That's the only one we have seen so far.

> Does this occur on all platforms or just MSVC ( maybe debug only )?

Correct. MSVC specific and on debug builds only.

> Our solutions will always be to move towards the C++ standard ...

I am not sure of what the standard says, but as per the [cppreference](https://en.cppreference.com/w/cpp/container/vector/operator_at), _Accessing a nonexistent element through this operator is undefined behavior._ May be replacing `m_data[0]` with `m_data.data()` would be an _acceptable_ solution since, again as per [cppreference](https://en.cppreference.com/w/cpp/container/vector/data) _For non-empty containers, the returned pointer compares equal to the address of the first element._ but wouldn't trigger the assertion.

```
T *data() {
   return &m_data[0]; // Change to m_data.data();
}

const T *data() const {
   return &m_data[0]; // Change to m_data.data();
}

const T *constData() const {
   return &m_data[0]; // Change to m_data.data();
}
```

**Comment by hs-apotell** [comment]
> replacing `m_data[0]` with `m_data.data()`

Just confirmed that making this change fixes the issue.

**Comment by agserm** [comment]
Thanks for testing on MSVC. We found the same resolution, currently testing to ensure it works properly on all of our supported platforms. (We have more than 15.)

**Comment by agserm** [comment]
This has been resolved in 45fbe7c, thanks for reporting.
